core/src/main/scala/org/apache/spark/internal/io/SparkHadoopWriter.scala

```diff
@@ -104,12 +104,12 @@ object SparkHadoopWriter extends Logging {
       jobTrackerId: String,
       commitJobId: Int,
       sparkPartitionId: Int,
-      sparkAttemptNumber: Int,
+      sparkTaskId: Long,
       committer: FileCommitProtocol,
       iterator: Iterator[(K, V)]): TaskCommitMessage = {
     // Set up a task.
     val taskContext = config.createTaskAttemptContext(
-      jobTrackerId, commitJobId, sparkPartitionId, sparkAttemptNumber)
+      jobTrackerId, commitJobId, sparkPartitionId, sparkTaskId.toInt)
```

Comment:
is it safe?

Reply:
task id is unique across the entire Spark application, which means we may have very large task id in a long-running micro-batch streaming application.

If we do need an int here, I'd suggest we combine `stageAttemptNumber` and `taskAttemptNumber` into a int, which is much less risky.(Spark won't have a lot of stage/task attempts)

Reply:
Streaming still generates separate jobs / stages for each batch, right?

In that case this should be fine; this would only be a problem if a single stage has enough tasks to cover all the integer space (4 billion tasks). That shouldn't be even possible since I doubt that you'd be able to have more than `Integer.MAX_VALUE` tasks (and even that is unlikely to ever happen).

I could use `abs` here (and in the sql code) to avoid a negative value (potentially avoiding weird file names).

Reply:
I don't follow, the task ids increment across jobs.  so if you have a very long running application that continues to start new jobs you could potentially run out.

Reply:
But what does "run out" mean?

If your task ID goes past `Int.MaxValue`, you'll start getting negative values here. Eventually you'll get to a long value that wraps back again to `0` when cast to an integer:

```
(2L + Int.MaxValue + Int.MaxValue).toInt
res2: Int = 0
```

So for this to "not work", which means you'd have a conflict where two tasks will generate the same output file name based on all these values (stage, task, partition, etc, etc), you need that situation to happen, which means you need about 4 billion tasks in the same stage for this to be a problem.

In other situations, you may get weird values because of the cast, but it should still work.

Reply:
Ah I see what you are saying, we just need to make sure it going negative doesn't cause any side affects or anything unpexected

Reply:
I commented before I saw this thread, but I think it is better to use the TID because that is already exposed in the UI so it is better for tracking between UI tasks and logs. The combined attempt number isn't used anywhere so this would introduce another number to identify a task. And anyway, shifting by 16 means that these grow huge anyway.